src/serialization.rs

```diff
@@ -210,9 +213,86 @@ pub(crate) fn i2osp<L: ArrayLength<u8>>(
     Ok(output)
 }
 
+/// Simplifies handling of [`serialize()`] output and implements [`Iterator`].
+pub(crate) struct Serialized<'a, L1: ArrayLength<u8>, L2: ArrayLength<u8>> {
+    octet: GenericArray<u8, L1>,
+    input: Input<'a, L2>,
+}
+
+enum Input<'a, L: ArrayLength<u8>> {
+    Owned(GenericArray<u8, L>),
+    Borrowed(&'a [u8]),
+}
+
+impl<'a, L1: ArrayLength<u8>, L2: ArrayLength<u8>> IntoIterator for &'a Serialized<'a, L1, L2> {
+    type Item = &'a [u8];
+
+    type IntoIter = IntoIter<&'a [u8], 2>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        IntoIter::new([
+            &self.octet,
+            match self.input {
+                Input::Owned(ref bytes) => bytes,
+                Input::Borrowed(bytes) => bytes,
+            },
+        ])
+    }
+}
+
 // Computes I2OSP(len(input), max_bytes) || input
-pub(crate) fn serialize<L: ArrayLength<u8>>(input: &[u8]) -> Result<Vec<u8>, InternalError> {
-    Ok([&i2osp::<L>(input.len())?, input].concat())
+pub(crate) fn serialize<L: ArrayLength<u8>>(
+    input: &[u8],
+) -> Result<Serialized<L, U0>, InternalError> {
+    Ok(Serialized {
+        octet: i2osp::<L>(input.len())?,
+        input: Input::Borrowed(input),
+    })
+}
+
+// Variation of `serialize` that takes an owned `input`
+pub(crate) fn serialize_owned<L1: ArrayLength<u8>, L2: ArrayLength<u8>>(
+    input: GenericArray<u8, L2>,
+) -> Result<Serialized<'static, L1, L2>, InternalError> {
+    Ok(Serialized {
+        octet: i2osp::<L1>(input.len())?,
+        input: Input::Owned(input),
+    })
+}
+
+macro_rules! chain_name {
+    ($var:ident, $mod:ident) => {
+        $mod
+    };
+    ($var:ident) => {
+        $var
+    };
+}
+
+macro_rules! chain_skip {
+    ($var:ident, $feed:expr) => {
+        $feed
+    };
+    ($var:ident) => {
+        $var
+    };
+}
+
+/// The purpose of this macro is to simplify [`concat`](alloc::slice::Concat::concat)ing
+/// slices into an [`Iterator`] to avoid allocation
+macro_rules! chain {
```

Comment:
I'm not sure this belongs here, I was thinking of making a `util` module and moving `i2osp`, `serialize`, `Serialized` and this macro into it. Any thoughts?

Reply:
I agree, that would be cleaner!

Reply:
(Will be happy to merge once this is moved)

Reply:
Done.